### ISSUE TITLE
Support arithmetic parenthesis

### ISSUE THREAD

**Issue by jgrenat** [systemjs/builder#286]
Hi!

My suggestion is quite simple. First, just have a look to this modules tree:

```
src/
    news/
        news-module1.js
        news-module2.js
    forum/
        forum-module1.js
        forum-module2.js
    startup.js
```

Let's say I want to bundle news files into a news.js file and everything else in a global.js file. I will do something like:

```bash
jspm bundle "startup - news/*" src/global.js --inject
jspm bundle "news/* - global" src/news.js --inject
```

If I use a dependency like *lodash* into my news bundle and somewhere else in the project, it won't be included into *global.js* and will be present into *news.js*. Of course, I could correct it manually in my bundle config, but the point is to automatize it.

Is it possible to have a new operator like ~ that would work like this (and would you find this interesting?):

```
# Load startup and its dependencies, except files in the news folder and their dependencies 
# that are not used anywhere else in startup
jspm bundle "startup ~ news/*" src/global.js --inject
```

## Alternative solution

Another solution (maybe simpler) would be to handle parenthesis in module expressions, because we could just do:

```
jspm bundle "startup - news/* + (startup & news/*)" src/global.js --inject
```

What do you think?

jgrenat

**Comment by guybedford** [comment]
If you create a bundle containing `startup & news/*` you can then use that bundle in subsequent subtractions.

Or would you never actually have a bundle containing `startup & news/&` here?

Supporting parenthesis is something that's always been an option, I think that may be the best approach here.

**Comment by jgrenat** [comment]
That's the point, I don't want a bundle containing `startup & news/*`.

I agree with the fact that supporting parenthesis would be a better option there ; for what I've seen, it will imply quite a lot of work. 
I'm not sure to have the time to look at it these days, but I will try to start something this week-end!

**Comment by guybedford** [comment]
Sure, yes it would require a little bit of consideration to implement, but shouldn't be too much code actually.

**Comment by rj33** [comment]
I'd also like to see parenthesis support, I think my use case is probably similar, I'd like to do:

(module1.js & module2.js) + (module3.js & module4.js) => common.js

Then 
(module1.js + module2.js) - common.js => bundle1.js
(module3.js + module4.js) - common.js => bundle2.js

From what I can see right now, this is probably only possible by creating intermediate bundles, although maybe I'm missing something in the way arithmetic operator precedence works that would let me achieve this already? 

Looking at the code it seems to have a simple left to right operator application, meaning mixing + and - would be ok, but not & and +, so the bundle1.js and bundle2.js would be fine without parenthesis, but the common.js case would be problematic without creating intermediate bundles that I don't need in the final outcome.  

Basically I'd like to avoid the cost of writing out intermediate bundles I don't intend t use - I notice in memory building is supported, but using tools like Grunt can make it trickier to make use of those abstractions in the underlying builder.

**Comment by guybedford** [comment]
Yes exactly this is not currently possible without intermediate bundles.

As a first step here we could potentially look at simply introducing operator precedence of `&` over `+` and `-`?

Contributions are also welcome along these lines.

**Comment by arackaf** [comment]
I might be up for implementing parenthesis support if you're interested.  I think the following would be a relatively straightforward implementation:

Modify parseExpression to be smart enough to spot parens.  When it does, it adds an entry to operations with the operator set to, say, `()` and a property of, say, `operations` which is an **array** set to a recursive call to parseExpression, passed the inner text of the parens content.

In traceExpression, refactor this code

  operations.forEach(function(operation) {
    expandPromise = expandPromise.then(function() {
      return Promise.resolve(expandGlobAndCanonicalize(builder, operation))
      .then(function(expanded) {
        expandedOperations = expandedOperations.concat(expanded);
      });
    });
  });

into a helper function which would take the current expandPromise, and an array of operations.  It would then recursively call itself when it comes across a parens operator from the previous step: it would pass the same current expandPromise, and of course the operations array added to the current entry by parseExpression in the prior step.

Would that work?  Am I missing any edge cases?  If you'd be interested in a PR I'd be happy to give it a go.

**Comment by guybedford** [comment]
@arackaf that would be amazing. The most important thing from a parsing perspective is that we maintain a carefully defined meaning of what characters can and can't be used in module names. Currently the parsing rule is simply to split by a space character, assuming that every space is followed by an operator, another space, then a module. When parsing brackets, for example `A + (B + C)` might make sense in this same model since we detect the paren "expression" after space parsing that it starts and ends in a bracket, and can still use spacing as the primary parsing concept. Let me know if that makes sense. From a code perspective as long as it is readable and thought out, I'm happy.

**Comment by arackaf** [comment]
@guybedford sweet - so the proposed implementation above would likely work?

Anything I need to know before diving into this?  Looks like you have tabs set to 2 spaces...any other styles you care about??

**Comment by guybedford** [comment]
@arackaf great. The main thing I care about is having as little code to maintain as possible. The clearer and smaller the code, the better.

**Comment by arackaf** [comment]
We think alike :)

I'll see what I can do.

**Comment by guybedford** [comment]
The implementation approach sounds great. I'll still be offline for the next couple of days unfortunately, but just let me know if you hit any edge cases you'd like to discuss and I'll do my best to respond.

**Comment by arackaf** [comment]
Much appreciated.  I assume you want to keep things as simple as possible?  Ie, I'd be happy to set you up with a Babel process, but then again might be more work than it's worth.

**Comment by guybedford** [comment]
@arackaf yeah I know it seems somewhat contradictory, but happy to write to the baseline directly.

**Comment by arackaf** [comment]
Not at all - one fewer thing to maintain.

**Comment by guybedford** [comment]
This was implemented by @arackaf already...
